### ISSUE TITLE
SetValidator inherited rulesets behavior

### ISSUE THREAD

**Issue by CesarD** [FluentValidation/FluentValidation#1504]
Documentation explains the following:

> Validating a specific ruleset with SetValidator
> Previously, if you defined a child validator with `SetValidator`, then whichever ruleset you invoked on the parent validator will cascade to the child validator. Now you can explicitly define which ruleset will run on the child:
> 
> `RuleFor(x => x.Address).SetValidator(new AddressValidator(), "myRuleset");`
> 

From what I've just experienced, if I don't explicitly set a ruleset when calling `SetValidator()`, the newly created validator (like `AddressValidator` in that example), will not contemplate any ruleset at all... 
This poses me a problem, because the previous behavior that "automatically inherited" the parent validator ruleset allowed me to not having to worry about passing any extra stuff if it wasn't any different... How can I achieve the same functionality? As far as I can see, I can't pass the current ruleset that my validator got passed from above so any new validator I create inside it, infers or inherits the same.

**Comment by JeremySkinner** [comment]
> he previous behavior that "automatically inherited" the parent validator ruleset allowed me to not having to worry about passing any extra stuff if it wasn't any different... How can I achieve the same functionality?

This behaviour is still in place - it hasn't changed. Whichever ruleset you execute on the root level will cascade to the child validator, but you must ensure that the parent rule is inside the same ruleset as the child rules. This behaviour hasn't changed. 

```csharp
public PersonValidator() {
  RuleSet("MyRuleSet", () => {
     RuleFor(x => x.Address).SetValidator(new AddressValidator());
  });
}

public AddressValidator() {
  RuleSet("MyRuleSet", () => {
     RuleFor(x => x.Line1).NotNull();
  });
} 
```

Note that the call to `SetValidator` must be wrapped in the ruleset you want to execute, or it won't run. Choosing to run a specific ruleset applies to *all validators in the hierarcy*. If the PersonValidator doesn't define "MyRuleSet", then nothing will execute.

**Comment by CesarD** [comment]
Thanks for such a fast response, Jeremy.

Ok, I get what you mean.

And what about this scenario?: 

```
public Validator1() {
   RuleFor(x => x.Field1).SetValidator(new Validator2(), "RuleSet1");
}

public Validator2() {
  RuleSet("RuleSet1", () => {
     RuleFor(x => x.Field1).NotNull();
  });

  RuleForEach(x => x.Collection1)
     .SetValidator(new Validator3());
} 

public Validator3() {
  RuleFor(x => x.Field1).NotNull();

  RuleSet("RuleSet1", () => {
    RuleFor(x => x.Field2).NotNull();
  });
}

```

In this case I would need to run some rules from `Validator3` always, and some only for a ruleset. But the case is, `Validator2` might need some rules ran within the ruleset, but the collection validation must always be executed, no matter what ruleset.
What would be the correct way to solve this behavior?

Thanks!

**Comment by JeremySkinner** [comment]
> but the collection validation must always be executed, no matter what ruleset.

If you wrap it in `RuleSet("*")` then it'll always be executed

**Comment by CesarD** [comment]
> > but the collection validation must always be executed, no matter what ruleset.
> 
> If you wrap it in `RuleSet("*")` then it'll always be executed

Wow, I didn't know that! 😄 
Is that in some part of the docs? I went over and over them this afternoon and never crossed it! 😅 
Thanks a lot, I will try it out

**Comment by JeremySkinner** [comment]
I’m not sure, it’s possible it hasn’t made it into the docs yet, I’ll try and get it added when I have some time.

**Comment by CesarD** [comment]
> > but the collection validation must always be executed, no matter what ruleset.
> 
> If you wrap it in `RuleSet("*")` then it'll always be executed

Sorry to bother you again, but what would be an example of how to wrap things with `RuleSet("*")` ?

Because I was trying the following:

```csharp

public Validator1() {
   RuleFor(x => x.Field1).SetValidator(new Validator2(), "RuleSet1");
}

public Validator2() {
  RuleSet("RuleSet1", () => {
     RuleFor(x => x.Field1).NotNull();
  });

  RuleFor(x => x.Field2).NotNull();

  RuleSet("*", () => {
    RuleForEach(x => x.Collection1)
       .SetValidator(new Validator3());
    }
  }
} 

public Validator3() {
  RuleFor(x => x.Field1).NotNull();

  RuleSet("RuleSet1", () => {
    RuleFor(x => x.Field2).NotNull();
  });
}
```

But the `Field2` validation from `Validator3` is not triggering.

**Comment by JeremySkinner** [comment]
Which ruleset are you invoking at the root level? "RuleSet1"?

**Comment by CesarD** [comment]
From my Controller, I'm invoking with no ruleset. I'm only dictating the rulesets for the validators I instance myself like in 

```csharp
public Validator1() {
   RuleFor(x => x.Field1).SetValidator(new Validator2(), "RuleSet1");
}
```

`Validator2` might be used somewhere else, but I want it to check `Field1` only when it's called with ruleset "RuleSet1". At the same time, I want `Validator3` to get the same behavior: will be set and run always, but some of its validations must run only when ruleset "RuleSet1" was set up above it.

**Comment by JeremySkinner** [comment]
> From my Controller, I'm invoking with no ruleset.

Well, that's why. The rule for `Field2` in Validator3 is explicitly wrapped in "RuleSet1", but you've not included RuleSet1 for execution. The ruleset cascade only works if you're actually triggering a ruleset. In your case you're not running a ruleset, so the process runs like this:

- No rulesets are included for execution, so only rules not in a ruleset will be run by default.
- Validator1.Rule1 is encountered. This rule is not in a ruleset, so it will be executed. 
  - Normally, only rules in Validator2 not in a ruleset will be run (as you haven't specified a ruleset to execute)
  - However, as Validator2 has been instantiated with "RuleSet1" in the call to SetValidator, RuleSet1 will be invoked *instead*.
    - Validator2.Field1 is inside "RuleSet1", so it will be executed
    - Validator2.Field2 is not inside a ruleset, so it won't be executed
    - Validator2.Field3 is inside "Ruleset *", so it will be executed
      - Validator3 has been instantiated *without a ruleset parameter*, so only rules not in a ruleset will run.
        - Validator3.Field1 is not in a ruleset, so it will execute
        - Validator3.Field3 is in ruleset "RuleSet1" so it *won't execute* 

If you want Validator3's "RuleSet1" to execute, you need to specify it in the call to SetValidator, in the same way you did for the root validator. The cascading/automatic propogation of selected rulesets only applies if you're actually executing a specific ruleset, which you're not.

```csharp
 RuleSet("*", () => {
    RuleForEach(x => x.Collection1)
       .SetValidator(new Validator3(), "RuleSet1");
    }
  }
```

If you do that, RuleSet1 will always be executed inside Validator3.

If you're still having trouble understanding, please put together some runnable unit tests and I will run them and show you what you need to change

**Comment by CesarD** [comment]
Thanks @JeremySkinner for the detailed explanation.
Sorry for the late reply but got into holidays and only today I'm catching up with this topic.

I guess part of my confusion comes from the fact that I'm setting the RuleSet on the child validators from the parent, but because my parent validator (Validator1) is expected to run only for a specific scenario, and then the children validators (Validator2 and Validator3) are more general ones, which can be run either on same scenario as Validator1 as well as by default on another ones.
If I explicitly set "RuleSet1" while instancing Validator3, it will always run under that ruleset when called from Validator2, no matter if Validator2 is being instanced by Validator1 or by some other validator (which it will, hence the problem).

Would be great if I could have some context data that indicates the current ruleset it's running on so I could pass it along when instancing validators from another one. Would that be something too weird or totally unnecessary?

**Comment by JeremySkinner** [comment]
> Would be great if I could have some context data that indicates the current ruleset it's running

This is already present inside the ValidationContext, but this is only available when validation is *executed*, not at the point when rules are defined, so this wouldn't really make sense. At the point the rules are defined, you haven't invoked validate yet so the validator has no idea what you're planning to execute. 

> If I explicitly set "RuleSet1" while instancing Validator3, it will always run under that ruleset when called from Validator2, no matter if Validator2 is being instanced by Validator1 or by some other validator (which it will, hence the problem).

That's correct, but you can just define the call to SetValidator multiple times if you need to, depending on the scenario.

However, personally I think your setup is too complicated and hard to follow using rulesets like this (as illustrated by this discussion). I'd suggest you consider splitting these into separate classes and then being very explicit about what you're running in each scenario. Ruleset cascade is *complicated* and it only gets more-so when you're trying to customise it at multiple levels. This feels like it's going to cause a huge maintenance headache if you continue down this route.

**Comment by CesarD** [comment]
Thank you very much for all your feedback, Jeremy.
